### PR TITLE
refactor tcp

### PR DESCRIPTION
**Pull request by turuslan**
- `SslContext` with p2p certificate for tls and quic
- read and resolve tcp/udp multiaddress
- fix `Multiaddress.hasProtocol` without value
- QUIC_V1 enum
- fix `Protocol` const prevented reassigning
- `Multiaddress` formatter
- fix `asioBuffer` ambigous when mutable

### CODE REVIEW

File: include/libp2p/transport/tcp/tcp_util.hpp [hunk 1]
```diff
@@ -6,134 +6,200 @@
...
-        if (protocol == P::WS) {
-          s << "/ws";
-        } else if (protocol == P::WSS) {
-          s << "/wss";
+      case P::DNS:
+      case P::DNS4:
+      case P::DNS6: {
+        Dns dns{std::nullopt, v};
+        switch (p.code) {
+          case P::DNS4:
```
Comment (igor-egorov) [inline include/libp2p/transport/tcp/tcp_util.hpp:52]: The case `P::DNS` is not covered by the inner switch.

Comment (igor-egorov) [reply include/libp2p/transport/tcp/tcp_util.hpp:52]: Probably it should be:
```suggestion
          case P::DNS:
          case P::DNS4:
```

**Review by igor-egorov — APPROVED** [review]
Please check the comments before merge.

File: include/libp2p/transport/tcp/tcp_util.hpp (hunk 1 above)
Comment (turuslan) [reply include/libp2p/transport/tcp/tcp_util.hpp:52]: No, `v4 = match { DNS => None, DNS4 => Some(true), DNS6 => Some(false) }`

**Pull request merged**